cpp/src/arrow/util/lazy-benchmark.cc

```diff
@@ -0,0 +1,123 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include <algorithm>
+#include <iterator>
+#include <vector>
+
+#include <benchmark/benchmark.h>
+
+#include "arrow/test-util.h"
+#include "arrow/util/lazy.h"
+
+namespace arrow {
+
+static constexpr int64_t kSize = 100000000;
+
+template <typename T = int32_t>
+std::vector<T> generate_junk(int64_t size) {
+  std::vector<T> v(size);
+  test::randint(size, 0, 100000, &v);
+  return v;
+}
+
+// baseline
+void BM_for_loop(benchmark::State& state) {
+  auto source = generate_junk(kSize);
+  std::vector<int> target(kSize);
+
+  for (auto _ : state) {
+    for (int64_t index = 0; index < kSize; ++index) target[index] = source[index] + 1;
+  }
+}
+
+BENCHMARK(BM_for_loop)->Repetitions(3)->Unit(benchmark::kMillisecond);
+
+// for comparison: pure copy without any changes
+void BM_std_copy(benchmark::State& state) {
+  auto source = generate_junk(kSize);
+  std::vector<int> target(kSize);
+
+  for (auto _ : state) {
+    std::copy(source.begin(), source.end(), target.begin());
+  }
+}
+
+BENCHMARK(BM_std_copy)->Repetitions(3)->Unit(benchmark::kMillisecond);
+
+// for comparison: pure copy without any changes
+void BM_std_copy_converting(benchmark::State& state) {
+  auto source = generate_junk<int32_t>(kSize);
+  // bigger type to avoid warnings
+  std::vector<int64_t> target(kSize);
+
+  for (auto _ : state) {
+    std::copy(source.begin(), source.end(), target.begin());
+  }
+}
+
+BENCHMARK(BM_std_copy_converting)->Repetitions(3)->Unit(benchmark::kMillisecond);
+
+// std::copy with a lazy iterator
+void BM_lazy_copy(benchmark::State& state) {
+  auto source = generate_junk(kSize);
+  std::vector<int> target(kSize);
+  auto lazy_range = internal::MakeLazyRange(
+      [&source](int64_t index) { return source[index]; }, source.size());
+
+  for (auto _ : state) {
+    std::copy(lazy_range.begin(), lazy_range.end(), target.begin());
+  }
+}
+
+BENCHMARK(BM_lazy_copy)->Repetitions(3)->Unit(benchmark::kMillisecond);
+
+// std::copy with a lazy iterator which does static cast
```

Comment:
I am curious: what do you call a lazy iterator? ISTM iterators are always lazy.

Reply:
I believe he's referring to the object returned by `internal::MakeLazyRange`

Reply:
Yes, the `LazyRange` is lazy about calling the supplied generator function instead of generating a new container eagerly like `std::generate`, for example. 

With these benchmarks I wanted to make sure, that the compilers can inline the lambdas properly so we don't sacrifice performance compared to a loop.